### PR TITLE
bnd 3.3.0 (new formula)

### PR DESCRIPTION
**Pull request by njbartlett**
Bnd is a popular Java tool for building, testing and
running OSGi Bundles. It is the engine behind a number
of other tools such as the Bndtools IDE (based on Eclipse)
along with various Maven plugins such as maven-bundle-plugin.
This formula enables bnd to be used as a standalone command-
line tool.

### CODE REVIEW

File: Formula/bnd.rb
Comment (woodruffw) [inline Formula/bnd.rb]: This is a good start, but something more complex than a simple `version`/`help` test is usually preferred. Is that possible with `bnd`?

Comment (bjhargrave) [inline Formula/bnd.rb]: What would you suggest? Bnd is already tested in its own build. There is no Bnd workspace known to the test to run larger tests upon.

Comment (woodruffw) [inline Formula/bnd.rb]: I'm not familiar with the tool, but perhaps constructing a simple workspace in the `test` sandbox and feeding it to `bnd`?

**Comment by njbartlett** [comment]
Its complicated because bnd works on binary artifacts: compiled java and
jars. I guess we would have to compile some simple java sources in the test
sandbox and then run against those.

Obviously the test cannot be comprehensive... Is it just intended as a
smoke test?

On 9 Dec 2016 4:30 p.m., "William Woodruff" <notifications@github.com>
wrote:

*@woodruffw* commented on this pull request.
------------------------------

In Formula/bnd.rb <https://github.com/Homebrew/homebrew-core/pull/7710>:

> +  url "https://search.maven.org/remotecontent?filepath=biz/aQute/bnd/biz.aQute.bnd/3.3.0/biz.aQute.bnd-3.3.0.jar"
+  sha256 "b6b68dfcd0f5ba767a202bf35eb3eb964c63e679e8217dd514dac807e6cedee8"
+
+  bottle :unneeded
+
+  def install
+    prefix.install "biz.aQute.bnd-#{version}.jar"
+
+    (bin/"bnd").write <<-EOS.undent
+      #!/bin/bash
+      exec java -jar #{prefix}/biz.aQute.bnd-#{version}.jar "$@"
+    EOS
+  end
+
+  test do
+    assert_match /^#{version}/, shell_output("#{bin}/bnd version")

I'm not familiar with the tool, but perhaps constructing a simple workspace
in the test sandbox and feeding it to bnd?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/Homebrew/homebrew-core/pull/7710>, or mute the thread

**Comment by woodruffw** [comment]
> Is it just intended as a smoke test?

Yep, just to make sure nothing explodes on first run.

Building a `jar` just to test might be overkill -- could it be run on its own jar (i.e., `biz.aQute.bnd-3.3.0.jar`)? If not, we can ignore my nit about `test`.

**Comment by njbartlett** [comment]
@woodruffw Thank you for your feedback, I have amended the test to perform a bnd resolution, which is one of the operations bnd performs on non-binary files.

**Comment by woodruffw** [comment]
Thanks!

**Comment by bjhargrave** [comment]
> could it be run on its own jar (i.e., biz.aQute.bnd-3.3.0.jar)?

You could just do `bnd print biz.aQute.bnd-3.3.0.jar` and sniff the output.

**Comment by woodruffw** [comment]
> You could just do bnd print biz.aQute.bnd-3.3.0.jar and sniff the output.

@njbartlett's added test should be sufficient, but this is good to know for future tests.

I'll let this sit for a bit to give the other maintainers a change to review, but after that I don't see any blocks on merging this.

File: Formula/bnd.rb
Comment (dunn) [inline Formula/bnd.rb]: We have a `#write_jar_script` helper method that should work here: 
```ruby
bin.write_jar_script prefix/"biz.aQute.bnd-#{version}.jar", "bnd"

**Comment by njbartlett** [comment]
@dunn Thank you I have made the suggested change.

File: Formula/bnd.rb
Comment (MikeMcQuaid) [inline Formula/bnd.rb]: Could simplify to `/launch.bndrun\s+ok/`

File: Formula/bnd.rb
Comment (MikeMcQuaid) [inline Formula/bnd.rb]: Could simplify to `/org.apache.felix.gogo.runtime.*org.apache.felix.gogo.runtime-1.0.0.jar/` (or better still using a variable)

File: Formula/bnd.rb
Comment (MikeMcQuaid) [inline Formula/bnd.rb]: Can remove this empty line.

File: Formula/bnd.rb
Comment (MikeMcQuaid) [inline Formula/bnd.rb]: Install to `libexec`

File: Formula/bnd.rb
Comment (MikeMcQuaid) [inline Formula/bnd.rb]: Can remove this comment (and perhaps this test)

File: Formula/bnd.rb
Comment (MikeMcQuaid) [inline Formula/bnd.rb]: can remove the second line of this comment

File: Formula/bnd.rb
Comment (MikeMcQuaid) [inline Formula/bnd.rb]: May want to consider making some of these valuables variables rather than hard-coding in two places.

File: Formula/bnd.rb
Comment (MikeMcQuaid) [inline Formula/bnd.rb]: Can remove this comment

File: Formula/bnd.rb
Comment (MikeMcQuaid) [inline Formula/bnd.rb]: again, could be a variable

File: Formula/bnd.rb
Comment (MikeMcQuaid) [inline Formula/bnd.rb]: Can remove this comment

File: Formula/bnd.rb
Comment (MikeMcQuaid) [inline Formula/bnd.rb]: Can remove this comment

File: Formula/bnd.rb
Comment (MikeMcQuaid) [inline Formula/bnd.rb]: Can remove this comment

**Review by MikeMcQuaid — CHANGES_REQUESTED** [review]

**Comment by njbartlett** [comment]
Changes requested by @MikeMcQuaid have been pushed.

**Comment by woodruffw** [comment]
Thanks @njbartlett. This LGTM, will wait a bit to see if anybody else has comments before merging.

**Comment by woodruffw** [comment]
Thanks for contributing, @njbartlett and @bjhargrave!

**Comment by njbartlett** [comment]
Awesome, thanks @woodruffw  and @MikeMcQuaid!

**Pull request closed without merging**
